[WebAssembly SIMD] Get WasmB3IRGenerator to parity on Intel

ddegazio opened the pull request:
#### 08944322951dfdb47c7a53273eb55cba2482d000
<pre>
[WebAssembly SIMD] Get WasmB3IRGenerator to parity on Intel
<a href="https://bugs.webkit.org/show_bug.cgi?id=249745">https://bugs.webkit.org/show_bug.cgi?id=249745</a>
rdar://103613145

Reviewed by Yusuke Suzuki.

Adds complete support for WebAssembly SIMD instructions to WasmB3IRGenerator. With this,
every WebAssembly SIMD instruction is supported on Intel in every JSC compiler tier. Because
of this, this patch also enables SIMD stress tests on x86_64 by default.

* Source/JavaScriptCore/b3/B3LowerToAir.cpp:
* Source/JavaScriptCore/wasm/WasmAirIRGenerator64.cpp:
(JSC::Wasm::AirIRGenerator64::addSIMDI_V):
(JSC::Wasm::AirIRGenerator64::addSIMDRelOp):
* Source/JavaScriptCore/wasm/WasmB3IRGenerator.cpp:
(JSC::Wasm::B3IRGenerator::addSIMDV_V):
(JSC::Wasm::B3IRGenerator::addSIMDRelOp):
(JSC::Wasm::B3IRGenerator::addSIMDSwizzleHelperX86):
(JSC::Wasm::B3IRGenerator::addSIMDV_VV):
(JSC::Wasm::B3IRGenerator::addSIMDExtmul):
(JSC::Wasm::B3IRGenerator::addSIMDShuffle):
* Tools/Scripts/run-jsc-stress-tests:

Canonical link: <a href="https://commits.webkit.org/258286@main">https://commits.webkit.org/258286@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1d07b3f23d28148ba51abbfb9dec7b78e93c9931

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/101446 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/10604 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/34500 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/110716 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/170972 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/105426 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/11556 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/1457 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/93833 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/108533 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/107228 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/8791 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/92031 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/35314 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/90689 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/23440 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/78314 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/91873 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/4206 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/24951 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/88019 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/1792 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/4268 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/1372 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/29319 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/10354 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/44439 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/90919 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5696 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/6031 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/20308 "Passed tests") | 
| | | | | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/b3/B3LowerToAir.cpp:
```diff
@@ -1291,9 +1291,17 @@ class LowerToAir {
-        append(isFloatingPoint ? Air::CompareFloatingPointVector : Air::CompareIntegerVector, 
-            cond, Arg::simdInfo(value->simdInfo()), tmp(value->child(0)), tmp(value->child(1)), tmp(value));
+        Air::Opcode airOp = isFloatingPoint ? Air::CompareFloatingPointVector : Air::CompareIntegerVector;
+        if (isValidForm(airOp, condKind, Arg::SIMDInfo, Arg::Tmp, Arg::Tmp, Arg::Tmp)) {
+            append(airOp,
+                cond, Arg::simdInfo(value->simdInfo()), tmp(value->child(0)), tmp(value->child(1)), tmp(value));
```
We would like to use early return if possible. Like,

```
if (isValidForm(airOp, condKind, Arg::SIMDInfo, Arg::Tmp, Arg::Tmp, Arg::Tmp)) {
    append(airOp, cond, Arg::simdInfo(value->simdInfo()), tmp(value->child(0)), tmp(value->child(1)), tmp(value));
    return;
}

if (isValidForm(airOp, condKind, Arg::SIMDInfo, Arg::Tmp, Arg::Tmp, Arg::Tmp, Arg::Tmp)) {
    ...;
    return;
}

RELEASE_ASSERT_NOT_REACHED();
```

ddegazio replied:
Refactored every case I could find to favor early return in the latest revision.

Constellation commented on Source/JavaScriptCore/b3/B3LowerToAir.cpp:
```diff
@@ -1331,7 +1339,13 @@ class LowerToAir {
+        else if (isValidForm(op, Arg::SIMDInfo, Arg::Tmp, Arg::Tmp, Arg::Tmp))
+            append(op, Arg::simdInfo(value->simdInfo()), tmp(value->child(0)), tmp(value), m_code.newTmp(FP));
+        else
+            RELEASE_ASSERT_NOT_REACHED();
```
Ditto.

Constellation commented on Source/JavaScriptCore/b3/B3LowerToAir.cpp:
```diff
@@ -3777,7 +3791,13 @@ class LowerToAir {
+                RELEASE_ASSERT(isX86());
+                append(op, scalar, tmp(value), m_code.newTmp(FP));
+            } else
+                RELEASE_ASSERT_NOT_REACHED();
```
Ditto.

Constellation commented on Source/JavaScriptCore/b3/B3LowerToAir.cpp:
```diff
@@ -3902,7 +3971,11 @@ class LowerToAir {
-            emitSIMDUnaryOp(Air::VectorAbs);
+            if (isX86() && m_value->as<SIMDValue>()->simdLane() == SIMDLane::i64x2) {
+                SIMDValue* value = m_value->as<SIMDValue>();
+                append(VectorAbsInt64, tmp(value->child(0)), tmp(value), m_code.newTmp(B3::FP));
```
Let's return early.

Constellation commented on Source/JavaScriptCore/b3/B3LowerToAir.cpp:
```diff
@@ -3920,13 +3993,46 @@ class LowerToAir {
+                    break;
+                default:
+                    RELEASE_ASSERT_NOT_REACHED();
+                }
```
Let's return early.

Constellation commented on Source/JavaScriptCore/b3/B3LowerToAir.cpp:
```diff
@@ -3920,13 +3993,46 @@ class LowerToAir {
-            emitSIMDUnaryOp(Air::VectorConvert);
+            if (isX86() && m_value->as<SIMDValue>()->signMode() == SIMDSignMode::Unsigned) {
+                SIMDValue* value = m_value->as<SIMDValue>();
+                append(VectorConvertUnsigned, tmp(value->child(0)), tmp(value), m_code.newTmp(B3::FP));
```
Ditto.

Constellation commented on Source/JavaScriptCore/b3/B3LowerToAir.cpp:
```diff
@@ -3920,13 +3993,46 @@ class LowerToAir {
+                if (value->signMode() == SIMDSignMode::Signed)
+                    append(VectorConvertLowSignedInt32, tmp(value->child(0)), tmp(value));
+                else
+                    append(VectorConvertLowUnsignedInt32, tmp(value->child(0)), tmp(value), m_code.newTmp(B3::GP), m_code.newTmp(B3::FP));
```
Ditto.

Constellation commented on Source/JavaScriptCore/b3/B3LowerToAir.cpp:
```diff
@@ -3967,10 +4073,21 @@ class LowerToAir {
+                if (value->simdLane() == SIMDLane::i16x8 && value->signMode() == SIMDSignMode::Unsigned)
+                    append(VectorExtaddPairwiseUnsignedInt16, tmp(value->child(0)), tmp(value), m_code.newTmp(B3::FP));
+                else
+                    append(Air::VectorExtaddPairwise, Arg::simdInfo(value->simdInfo()), tmp(value->child(0)), tmp(value), m_code.newTmp(B3::GP), m_code.newTmp(B3::FP));
```
Ditto.

Constellation commented on Source/JavaScriptCore/b3/B3LowerToAir.cpp:
```diff
@@ -3967,10 +4073,21 @@ class LowerToAir {
-            emitSIMDMonomorphicBinaryOp(Air::VectorMulSat);
+            if constexpr (isX86()) {
+                SIMDValue* value = m_value->as<SIMDValue>();
+                append(Air::VectorMulSat, tmp(value->child(0)), tmp(value->child(1)), tmp(value), m_code.newTmp(GP), m_code.newTmp(FP));
```
Ditto.

Constellation commented on Source/JavaScriptCore/wasm/WasmB3IRGenerator.cpp:
```diff
@@ -391,6 +391,80 @@ class B3IRGenerator {
         B3_OP_CASE(Neg)
+
+        if (isX86()) {
+            if (b3Op == VectorPopcnt) {
```
I think this should be implemented in `B3LowerMacros` so that all B3 frontend can use `VectorPopcnt`.

ddegazio replied:
Moved everything from the IR generator to `B3LowerMacros.cpp` in latest revision.

Constellation commented on Source/JavaScriptCore/wasm/WasmB3IRGenerator.cpp:
```diff
@@ -391,6 +391,80 @@ class B3IRGenerator {
+                return { };
+            }
+
+            if (b3Op == VectorNot) {
```
Ditto.

Constellation commented on Source/JavaScriptCore/wasm/WasmB3IRGenerator.cpp:
```diff
@@ -391,6 +391,80 @@ class B3IRGenerator {
+                return { };
+            }
+
+            if (b3Op == VectorNeg) {
```
Ditto.

Constellation commented on Source/JavaScriptCore/wasm/WasmB3IRGenerator.cpp:
```diff
@@ -440,6 +514,51 @@ class B3IRGenerator {
                 break;
             }
+
+            if constexpr (isX86()) {
```
Ditto.

Constellation commented on Source/JavaScriptCore/wasm/WasmB3IRGenerator.cpp:
```diff
@@ -2931,6 +3072,29 @@ auto B3IRGenerator::addSIMDExtmul(SIMDLaneOperation op, SIMDInfo info, Expressio
 
 auto B3IRGenerator::addSIMDShuffle(v128_t imm, ExpressionType a, ExpressionType b, ExpressionType& result) -> PartialResult
 {
+    if (isX86()) {
```
Ditto.

Constellation reviewed (commented):
Comments

Constellation commented on Source/JavaScriptCore/b3/B3LowerToAir.cpp:
```diff
@@ -3827,7 +3855,41 @@ class LowerToAir {
+                    // x86 has no SIMD 64-bit signed right shift instruction, so we scalarize it here.
+
+#if CPU(X86_64)
+                    Tmp shiftRCX = Tmp(X86Registers::ecx);
```
IIRC, we have m_ecx for this.

ddegazio replied:
Will change it to use `m_ecx`, that definitely sounds like the right choice.

ddegazio replied:
Updated.

justinmichaud commented on Source/JavaScriptCore/b3/B3LowerToAir.cpp:
```diff
@@ -3827,7 +3855,41 @@ class LowerToAir {
+                if (value->opcode() == VectorShr && value->signMode() == SIMDSignMode::Signed && value->simdLane() == SIMDLane::i64x2) {
+                    // x86 has no SIMD 64-bit signed right shift instruction, so we scalarize it here.
+
+#if CPU(X86_64)
```
This is already in an if?

ddegazio replied:
The first `if` doesn't protect from the fact we use `X86Registers::ecx`, which won't compile on non-Intel. So I guess we could make the whole thing part of the `#if CPU(X86_64)`.  Better would just be to get rid of it though, since this class also defines `m_ecx` which I could just use instead. I'll use that instead and remove the preprocessor directives.

ddegazio replied:
Updated.

Constellation commented on Source/JavaScriptCore/b3/B3LowerMacros.cpp:
```diff
@@ -417,14 +417,187 @@ class LowerMacros {
+                    break;
+                }
+                case SIMDLane::f64x2: {
+                    Value* topBit = m_insertionSet.insert<Const64Value>(m_index, m_origin, -0x8000000000000000);
```
Probably ULL postfix is necessary for this integer.

ddegazio replied:
Updated.

justinmichaud commented on Source/JavaScriptCore/b3/B3LowerToAir.cpp:
```diff
@@ -3796,8 +3816,16 @@ class LowerToAir {
+                return;
+            }
+            if (isValidForm(op, Arg::Tmp, Arg::Tmp, Arg::Tmp)) {
+                RELEASE_ASSERT(isX86());
```
Is this needed?

ddegazio replied:
Huh. Doesn't seem like we do, I'll make sure there wasn't some failure we ran into that provoked this though.

ddegazio replied:
Removed it and everything still works, as one would expect in hindsight. Good catch!

Constellation reviewed (approved):
r=me,aybe, uint64_t constants need ULL postfix.

justinmichaud commented on Source/JavaScriptCore/b3/B3LowerToAir.cpp:
```diff
@@ -3986,9 +4107,22 @@ class LowerToAir {
+            if (isX86()) {
+                SIMDValue* value = m_value->as<SIMDValue>();
+                if (value->simdLane() == SIMDLane::i16x8 && value->signMode() == SIMDSignMode::Unsigned)
+                    append(VectorExtaddPairwiseUnsignedInt16, tmp(value->child(0)), tmp(value), m_code.newTmp(B3::FP));
```
Why can't we just always emit VectorExtaddPairwise, and selectively lower in lowerMacros? We can make tmps there too.

ddegazio replied:
I just copied our implementation from `AirIRGenerator`, but good point! We can probably remove this from here and from Air and replace both with a lower macro.

ddegazio replied:
Updated.

ddegazio committed:
[WebAssembly SIMD] Get WasmB3IRGenerator to parity on Intel
https://bugs.webkit.org/show_bug.cgi?id=249745
rdar://103613145

Reviewed by Yusuke Suzuki.

Adds complete support for WebAssembly SIMD instructions to WasmB3IRGenerator. With this,
every WebAssembly SIMD instruction is supported on Intel in every JSC compiler tier. Because
of this, this patch also enables SIMD stress tests on x86_64 by default.

* Source/JavaScriptCore/b3/B3LowerToAir.cpp:
* Source/JavaScriptCore/wasm/WasmAirIRGenerator64.cpp:
(JSC::Wasm::AirIRGenerator64::addSIMDI_V):
(JSC::Wasm::AirIRGenerator64::addSIMDRelOp):
* Source/JavaScriptCore/wasm/WasmB3IRGenerator.cpp:
(JSC::Wasm::B3IRGenerator::addSIMDV_V):
(JSC::Wasm::B3IRGenerator::addSIMDRelOp):
(JSC::Wasm::B3IRGenerator::addSIMDSwizzleHelperX86):
(JSC::Wasm::B3IRGenerator::addSIMDV_VV):
(JSC::Wasm::B3IRGenerator::addSIMDExtmul):
(JSC::Wasm::B3IRGenerator::addSIMDShuffle):
* Tools/Scripts/run-jsc-stress-tests:

Canonical link: https://commits.webkit.org/258286@main
